File: Easy/JAVASCRIPT/ThreeLargestNumbers.js
const CAPACITY = 3;
const getLargestNumbers = (array) => {
    array.sort((a, b) => a - b);
    let subArray = [...Array(CAPACITY)];
    // let subArray = new Array(CAPACITY).fill(0);
    let j = 0;
    for (let i = 0; i < array.length; i++)
        if (i >= array.length - CAPACITY)
            subArray.splice(j++, 1, array[i]);

    return subArray;
}


const getThreeLargestNumbers = (array) => {
    const threeLargestNumbers = [null, null, null];
    for (const number of array)
        validate(threeLargestNumbers, number);

    return threeLargestNumbers;
}

const validate = (array, number) => {
    if (array[2] === null || number > array[2])
        update(array, number, 2);
    else if (array[1] === null || number > array[1])
        update(array, number, 1);
    else if (array[0] === null || number > array[0])
        update(array, number, 0);

}

const update = (array, number, index) => {
    for (let i = 0; i <= index; i++)
        if (i === index) array[i] = number;
        else array[i] = array[i + 1];
}




const numbers = [141, 1, 17, -7, -17, -27, 18, 541, 8, 7, 7];
// const result = getLargestNumbers(numbers);
const result = getThreeLargestNumbers(numbers);
console.log(result);

File: Easy/PYTHON/ThreeLargestNumbers.py

CAPACITY = 3
# subArray = list(None for i in range(0, CAPACITY))
subArray = [None]*CAPACITY


def get_largest_numbers(array):
    array.sort()
    j = 0
    for i in range(0, len(array)):
        if i >= len(array) - CAPACITY:
            subArray[j] = array[i]
            j += 1

    return subArray


def get_three_largest_numbers(array):
    threeLargestNumbers = subArray
    for number in array:
        validate(threeLargestNumbers, number)

    return threeLargestNumbers


def validate(array, number):
    if array[2] is None or number > array[2]:
        update(array, number, 2)
    elif array[1] is None or number > array[1]:
        update(array, number, 1)
    elif array[0] is None or number > array[0]:
        update(array, number, 0)


def update(array, number, index):
    for i in range(index +1):
        if i == index:
            array[i] = number
        else:
            array[i] = array[i + 1]


numbers = [141, 1, 17, -7, -17, -27, 18, 541, 8, 7, 7]
# result = get_largest_numbers(numbers)
result = get_three_largest_numbers(numbers)
print(result)
